### PR TITLE
[v7r0] refactor documentation scripts to be used in extensions

### PR DESCRIPTION
**Pull request by andresailer**
This is a refactoring of the scripts creating the

- [x] rst files for code documentation
- [x] command references. 
- [x] CFG concat
- [x] release notes
- [x] cleanup
- [x] consistent variable names

After this refactoring they can be simply imported into extension's `conf.py`, configured via a `docs.conf` file, or run manually.
See here for now: https://diracsailer.readthedocs.io/en/fixdoc/DeveloperGuide/CodeDocumenting/BuildingDocumentation.html

BEGINRELEASENOTES

*Docs
CHANGE: refactor existing Tools into diracdoctools and make them usable by extension
CHANGE: GetReleaseNotes.py is now `dirac-docs-get-release-notes` and also works on gitlab, see its `--help` message. Tested on gitlab.cern.ch, but I think it should also work on other gitlab instances, as long as there is the API v4 enabled

ENDRELEASENOTES

### CODE REVIEW

**Comment by andresailer** [comment]
Review please

**Comment by chaen** [comment]
you maybe just want to squash all the commit before merging ? :D

File: docs/diracdoctools/__init__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+"""Tools to build the DIRAC documentation.
+
+
+Scripts to build code documentation, and also some mocking, etc.
+"""
+
+# list of packages that should be mocked in sphinx
+# TODO: check if we can install more of these
+DIRAC_DOC_MOCK_LIST = ['XRootD',
```
[the remark this answers is not in the document]
Comment (andresailer) [reply docs/diracdoctools/__init__.py]: Let's see, removed it

**Comment by andresailer** [comment]
There is some information about what things do not work in the commit messages

File: docs/Makefile [hunk 2]
```diff
@@ -46,6 +47,17 @@ html:
 	@echo
 	@echo "Build finished. The HTML pages are in $(BUILDDIR)/html."
 
+# AUTO MAKE START
+allhtml:
```
Comment (andresailer) [inline docs/Makefile]:
```suggestion
htmlall:
```

File: docs/Makefile [hunk 3]
```diff
@@ -28,6 +28,7 @@ ALLSPHINXOPTS   = -d $(BUILDDIR)/doctrees $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) sou
 help:
 	@echo "Please use \`make <target>' where <target> is one of"
 	@echo "  html      to make standalone HTML files"
+	@echo "  allhtml   to make standalone HTML files"
```
Comment (andresailer) [inline docs/Makefile]:
```suggestion
	@echo "  htmlall   to make standalone HTML files, including auto generating various source files"
```

File: docs/source/DeveloperGuide/CodeDocumenting/BuildingDocumentation.rst [hunk 4]
```diff
@@ -0,0 +1,94 @@
+.. _building_documentation:
+
+==========================
+Building the Documentation
+==========================
+
+The DIRAC documentation is crated using ``sphinx`` and makes use of the
```
Comment (andresailer) [inline docs/source/DeveloperGuide/CodeDocumenting/BuildingDocumentation.rst]:
```suggestion
The DIRAC documentation is created using ``sphinx`` and makes use of the
```

**Comment by andresailer** [comment]
From my point of view, this PR is ready to be merged, failing test for docs due to things fixed in #4029

File: docs/diracdoctools/Config.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
+"""Configuration for the documentation scripts."""
+
+import argparse
+from collections import defaultdict
+import logging
+import os
+from pprint import pformat
+import ConfigParser
```
Comment (fstagni) [inline docs/diracdoctools/Config.py]: better order?

File: docs/diracdoctools/Config.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+K_COMMANDS = 'Commands'
+K_CFG = 'CFG'
+
+
+def listify(values):
+  return [entry.strip() for entry in values.split(',') if entry]
+
+
+class Configuration(object):
+  """Provide configuraiton to the scripts."""
```
Comment (fstagni) [inline docs/diracdoctools/Config.py]:
```suggestion
  """Provide configuration to the scripts."""
```

File: docs/source/DeveloperGuide/CodeDocumenting/BuildingDocumentation.rst [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+then be called to create code reference, command reference and concatenated CFG
+files.
+
+.. literalinclude:: ../../conf.py
+   :start-after: # AUTO SETUP START
+   :end-before:  # AUTO SETUP END
+   :caption: docs/source/conf.py
+
+
+The configuration for ``diracdoctools`` for is done via a configuration file
```
Comment (fstagni) [inline docs/source/DeveloperGuide/CodeDocumenting/BuildingDocumentation.rst]:
```suggestion
The configuration for ``diracdoctools`` is done via a configuration file
```

File: docs/requirements.txt [hunk 8]
```diff
@@ -9,3 +12,6 @@ pytz
 psutil
 pyparsing
 GSI
+M2Crypto
```
Comment (fstagni) [inline docs/requirements.txt]: Can we removed GSI?

Comment (andresailer) [reply docs/requirements.txt]: Yes, GSI dropped, though there is an issue with installing m2crypto 0.33 (released an hour ago) on RTD, so fixed to 0.32

Comment (andresailer) [reply docs/requirements.txt]: well, m2crypto 0.33 also doesnt work on travis

File: docs/diracdoctools/Config.py (hunk 5 above)
Comment (andresailer) [reply docs/diracdoctools/Config.py]: cleaned the imports

**Comment by andresailer** [comment]
I am sure there will be a world of unicode pain waiting when using python3

**Comment by andresailer** [comment]
What are the chances of this getting merged before the workshop?

**Comment by chaen** [comment]
they should be max :)

**Pull request merged**